store/tikv/lock_resolver.go

```diff
@@ -578,10 +590,149 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+		if cmdResp.UseAsyncCommit {
+			status.secondaries = cmdResp.GetSecondaries()
+		}
+
 		return status, nil
 	}
 }
 
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, lite bool, cleanRegions map[RegionVerID]struct{}) error {
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	commitTs := l.MinCommitTS
+	commitTsMutex := sync.Mutex{}
+	locks := []*Lock{l}
+	missingLock := false
+	errChan := make(chan error, len(regions))
+
+	for regionID, keys := range regions {
+		curRegionID := regionID
+		curKeys := keys
+
+		go func() {
+			for {
+				checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+					Keys:         curKeys,
+					StartVersion: l.TxnID,
+				}
+				req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+				tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+				resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				regionErr, err := resp.GetRegionError()
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				if regionErr != nil {
+					err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+					if err != nil {
+						errChan <- errors.Trace(err)
+						return
+					}
+					continue
+				}
+				if resp.Resp == nil {
+					errChan <- errors.Trace(ErrBodyMissing)
+					return
+				}
+				checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+
+				// Check each lock.
+				for _, lockInfo := range checkResp.Locks {
+					if lockInfo == nil {
```

Comment:
When one of the locks is missing, may it be nil? I'm afraid protobuf doesn't behaves like this. Neither will TiKV return a list of locks if one of them is missing. @sticnarf  PTAL.

Reply:
Here seems not right to me. I remember a protobuf repeated field cannot contain nil fields...

Reply:
We should check `len(checkResp.Locks)`. Currently if any lock does not exist, no lock is returned back.